cranelift/codegen/src/opts/div_const.rs

```diff
@@ -0,0 +1,1148 @@
+        for x in 0..(200 * 1000i64) {
+            let m = magic_s64(-0x8000_0000_0000_0000i64 + x);
+            total = total ^ (m.mul_by as u64);
+            total = total + (m.shift_by as u64);
+        }
+        assert_eq!(total, 17929885647724831014);
+
+        total = 0;
+        // Testing MIDPOINT magic_s64
+        for x in 0..(200 * 1000i64) {
+            let x2 = -100 * 1000i64 + x;
+            if x2 != -1 && x2 != 0 && x2 != 1 {
+                let m = magic_s64(x2);
+                total = total ^ (m.mul_by as u64);
+                total = total + (m.shift_by as u64);
+            }
+        }
+        assert_eq!(total, 18106042338125661964);
+
+        // Testing DOWN magic_s64
+        total = 0;
+        for x in 0..(200 * 1000i64) {
+            let m = magic_s64(0x7FFF_FFFF_FFFF_FFFFi64 - x);
+            total = total ^ (m.mul_by as u64);
+            total = total + (m.shift_by as u64);
+        }
+        assert_eq!(total, 563301797155560970);
+    }
+
+    // These generate reference results for signed/unsigned 32/64 bit
+    // division, rounding towards zero.
+    fn div_u32(x: u32, y: u32) -> u32 {
+        return x / y;
+    }
+    fn div_s32(x: i32, y: i32) -> i32 {
+        return x / y;
+    }
+    fn div_u64(x: u64, y: u64) -> u64 {
+        return x / y;
+    }
+    fn div_s64(x: i64, y: i64) -> i64 {
+        return x / y;
+    }
+
+    // Returns the high half of a 32 bit unsigned widening multiply.
+    fn mulhw_u32(x: u32, y: u32) -> u32 {
+        let x64: u64 = x as u64;
+        let y64: u64 = y as u64;
+        let r64: u64 = x64 * y64;
+        (r64 >> 32) as u32
+    }
+
+    // Returns the high half of a 32 bit signed widening multiply.
+    fn mulhw_s32(x: i32, y: i32) -> i32 {
+        let x64: i64 = x as i64;
+        let y64: i64 = y as i64;
+        let r64: i64 = x64 * y64;
+        (r64 >> 32) as i32
+    }
+
+    // Returns the high half of a 64 bit unsigned widening multiply.
+    fn mulhw_u64(x: u64, y: u64) -> u64 {
+        let t0: u64 = x & 0xffffffffu64;
+        let t1: u64 = x >> 32;
+        let t2: u64 = y & 0xffffffffu64;
+        let t3: u64 = y >> 32;
+        let t4: u64 = t0 * t2;
+        let t5: u64 = t1 * t2 + (t4 >> 32);
+        let t6: u64 = t5 & 0xffffffffu64;
+        let t7: u64 = t5 >> 32;
+        let t8: u64 = t0 * t3 + t6;
+        let t9: u64 = t1 * t3 + t7 + (t8 >> 32);
+        t9
+    }
+
+    // Returns the high half of a 64 bit signed widening multiply.
+    fn mulhw_s64(x: i64, y: i64) -> i64 {
+        let t0: u64 = x as u64 & 0xffffffffu64;
+        let t1: i64 = x >> 32;
+        let t2: u64 = y as u64 & 0xffffffffu64;
+        let t3: i64 = y >> 32;
+        let t4: u64 = t0 * t2;
+        let t5: i64 = t1 * t2 as i64 + (t4 >> 32) as i64;
+        let t6: u64 = t5 as u64 & 0xffffffffu64;
+        let t7: i64 = t5 >> 32;
+        let t8: i64 = t0 as i64 * t3 + t6 as i64;
+        let t9: i64 = t1 * t3 + t7 + (t8 >> 32);
+        t9
+    }
```

Comment:
FWIW while these were probably copied from C or similar you can probably simplify these greatly with 128-bit integers

Reply:
Also this sort of makes me wonder if we could have a monster ISLE rule to pattern match this and generate `smulhi`....

Reply:
The tricky bit with matching this gunk to produce `smulhi` is all of the commutative operations.

Reply:
I think what you're describing is the draft work I did, and comments I wrote, in https://github.com/bytecodealliance/wasmtime/pull/8653; I was trying to get all the way to wide imul, but I think similar reasoning should work for *mulhi.